typing/typedecl.ml

```diff
@@ -646,7 +646,134 @@ let check_coherence env loc dpath decl =
 let check_abbrev env sdecl (id, decl) =
   check_coherence env sdecl.ptype_loc (Path.Pident id) decl
 
-(* Check that recursion is well-founded:
+
+(* Note: Well-foundedness for OCaml types
+
+   We want to guarantee that all cycles within OCaml types are
+   "guarded".
+
+   More precisly, we consider a reachability relation
+     "[t] is reachable [guarded|unguarded] from [u]"
+   defined as follows:
+
+   - [t1, t2...] are reachable guarded from object types
+       [< m1 : t1; m2 : t2; ... >]
+     or polymorphic variants
+       [[`A of t1 | `B of t2 | ...]].
+
+   - [t1, t2...] are reachable rectypes-guarded from
+     [t1 -> t2], [t1 * t2 * ...], and all other built-in
+     contractive type constructors.
```

Comment:
Not only built-in, variant and record type constructors are also contractive and rectypes-guarded.

Reply:
With this relation as defined, the definition of a datatype is not reachable at all from instances of the datatype constructor. This is pointed out explicitly in a comment further down in the definition.

Reading this again, there is not much of a difference between "reachable guarded" and "not reachable at all". I guess I could provide a different specification where the definition of a datatype is reachable guarded, or where there is no guarded/unguarded distinction but just reachable types (corresponding to reachable-ungarded).

I'm not sure what would be clearer. Specifying only the reachable-ungarded part is more concise, but maybe the user expects to find any case that they wonder about explicitly defined as either "reachable" or "not reachable".

Reply:
The definition is not reachable, but the parameters are.
I.e. a datatype behaves exactly as a contractive type constructors for reachability.

Reply:
> The definition is not reachable, but the parameters are.

Yes, I believe this is already documented later on:

```
   - If [(t1, t2...) c] is not an expandable abbreviation,
     typically an abstract type or a datatype (variant or record)
     then [t1, t2...] are reachable unguarded from it.
```

> a datatype behaves exactly as a contractive type constructor for reachability.

If I understand correctly, an issue with the documentation extract listed above is that while datatypes have rectypes-guarded parameters, abstract types have unguarded parameters. Do you agree?

> Note that reachable-guarded is not completely irrelevant.

Good point, thanks! I guess it validates my intuition to make a distinction in the specification, even if we don't discuss these details.

Reply:
Yes, datatypes are necessarily contractive so they behave like builtin-types, but abstract types might be non-contractive so their parameters are unguarded.
(A point of the nominal type RFC is that nominal types would be both abstract and contractive.)

Reply:
I changed the documentation accordingly, thanks.